client/state/sites/selectors.js

```diff
@@ -68,36 +67,41 @@ export const getSiteBySlug = createSelector(
 );
 
 /**
- * Returns a normalized site object by its ID. Intends to replicate
- * the site object returned from the legacy `sites-list` module.
- *
+ * Returns a normalized site object by its ID or site slug.
  *
  * @param  {Object}  state  Global state tree
- * @param  {Number}  siteId Site ID
+ * @param  {Number}  siteId Site ID or site slug
  * @return {?Object}        Site object
  */
-export const getSite = createSelector(
-	( state, siteId ) => {
-		let site =
-			getRawSite( state, siteId ) ||
-			// Support for non-ID site retrieval
-			// Replaces SitesList#getSite
-			getSiteBySlug( state, siteId );
-
-		if ( ! site ) {
-			return null;
-		}
+let getSiteCache = new WeakMap();
 
-		// To avoid mutating the original site object, create a shallow clone
-		// before assigning computed properties
-		site = { ...site };
-		assign( site, getSiteComputedAttributes( state, siteId ) );
-		assign( site, getJetpackComputedAttributes( state, siteId ) );
+export function getSite( state, siteId ) {
+	const rawSite = getRawSite( state, siteId ) || getSiteBySlug( state, siteId );
+	if ( ! rawSite ) {
+		return null;
+	}
 
-		return site;
-	},
-	state => [ state.sites.items, state.currentUser.capabilities ]
-);
+	// Use the rawSite object itself as a WeakMap key
+	const cachedSite = getSiteCache.get( rawSite );
+	if ( cachedSite ) {
+		return cachedSite;
+	}
+
+	const site = {
+		...rawSite,
+		...getSiteComputedAttributes( state, siteId ),
+		...getJetpackComputedAttributes( state, siteId ),
+	};
+
+	// Once the `rawSite` object becomes outdated, i.e., state gets updated with a newer version
+	// and no more references are held, the key will be automatically removed from the WeakMap.
+	getSiteCache.set( rawSite, site );
```

Comment:
So tracing this down the `state.sites.items` array holds a hard reference to our `rawSite`s. So in theory we should be able to GC weak refs in the map after state.sites.items is updated again, right?

Reply:
the `rawSite` will be collected but the cache of it won't. as long as the actual `rawSite` reference is still in memory (such as in `state.sites.items`) then we'll be able to retrieve the value here. once it's no longer in use outside of this place it can be freed. at that point, we'll love our ability to retrieve the cached value; it's not clear to me if that value will leak or if the engine will free it as well

<img width="982" alt="screen shot 2017-11-15 at 4 20 36 pm" src="https://user-images.githubusercontent.com/5431237/32860927-2a1e5d6c-ca21-11e7-8ed0-c5370c26163b.png">

clearly I don't know what I'm doing here, but I tried to see if the weak map freed its local objects. in this ill-performed experiment i'm not sure what it indicates. but, the data is there for any more informed eyes

Reply:
The `WeakMap` entry holds a weak reference to the key and a strong reference to the value. If anyone is holding a strong reference to the key, they can use it to access the value in the map. 

If a reference to the key no longer exists, the entry becomes inaccessible and is removed from the map during the GC. During that GC, the strong reference to the value is removed. If it was the last reference to it, the value also becomes eligible for GC.

Reply:
> During that GC, the strong reference to the value is removed. If it was the last reference to it, the value also becomes eligible for GC.

This is what seems most expected to me, but I couldn't find the documentation for it; seems like it should be obvious.